base_layer/core/src/proof_of_work/monero_rx.rs

```diff
@@ -231,25 +231,343 @@ fn verify_root(monero_data: &MoneroData) -> Result<(), MergeMineError> {
+        let mut hashes = Vec::with_capacity(count as usize);
+        hashes.push(block.miner_tx.hash());
+        for item in block.clone().tx_hashes {
+            hashes.push(item);
+        }
+        let root = tree_hash(hashes.clone());
+        let input_blob = create_input_blob(&block.header, &count, &from_hashes(&hashes)).unwrap();
+        assert_eq!(input_blob, "0c0c8cd6a0fa057fe21d764e7abf004e975396a2160773b93712bf6118c3b4959ddd8ee0f76aad0000000058b030b6800d433bbcb2b560afe2a08e4dc152fa77ead96d37aaf14897d3c09601");
+    }
+
+    #[test]
+    fn test_append_mm_tag() {
+        let mut blocktemplate_blob = "0c0c8cd6a0fa057fe21d764e7abf004e975396a2160773b93712bf6118c3b4959ddd8ee0f76aad0000000002e1ea2701ffa5ea2701d5a299e2abb002028eb3066ced1b2cc82ea046f3716a48e9ae37144057d5fb48a97f941225a1957b2b0106225b7ec0a6544d8da39abe68d8bd82619b4a7c5bdae89c3783b256a8fa47820208f63aa86d2e857f070000".to_string();
+        let seed_hash = "9f02e032f9b15d2aded991e0f68cc3c3427270b568b782e55fbd269ead0bad97".to_string();
+        let bytes = hex::decode(blocktemplate_blob).unwrap();
+        let mut block = deserialize::<MoneroBlock>(&bytes[..]).unwrap();
+        let mut block_header = BlockHeader {
+            version: 0,
+            height: 0,
+            prev_hash: vec![0],
+            timestamp: Default::default(),
+            output_mr: vec![0],
+            range_proof_mr: vec![0],
+            kernel_mr: vec![0],
+            total_kernel_offset: RistrettoSecretKey::from(0),
+            nonce: 0,
+            pow: ProofOfWork::default(),
+        };
+        let hash = Hash::from(from_slice(&block_header.merged_mining_hash()));
+        blocktemplate_blob = append_merge_mining_tag(&mut block, hash).unwrap();
+        let count = 1 + (block.tx_hashes.len() as u16);
+        let mut hashes = Vec::with_capacity(count as usize);
+        let mut proof = Vec::with_capacity(count as usize);
+        hashes.push(block.miner_tx.hash());
+        proof.push(block.miner_tx.hash());
+        for item in block.clone().tx_hashes {
+            hashes.push(item);
+            proof.push(item);
+        }
+        let root = tree_hash(hashes.clone());
+        let monero_data = MoneroData {
+            header: block.header,
+            key: seed_hash,
+            count,
+            transaction_root: from_slice(&root),
+            transaction_hashes: from_hashes(&hashes),
+            coinbase_tx: block.miner_tx,
+        };
+        let serialized = bincode::serialize(&monero_data).unwrap();
+        let pow = ProofOfWork {
+            accumulated_monero_difficulty: Default::default(),
+            accumulated_blake_difficulty: Default::default(),
+            target_difficulty: Default::default(),
+            pow_algo: PowAlgorithm::Monero,
+            pow_data: serialized,
+        };
+        block_header.pow = pow;
+        let monero_data = MoneroData::new(&block_header).unwrap();
+        verify_header(&block_header, &monero_data).unwrap();
```

Comment:
This checks the happy path - but I think some tests that check the failure behaviour - i.e given inputs `bad_block_header` and `monero_data` I expect error X etc. are valuable.

Reply:
Fair comment, will add more tests. Thanks.

Reply:
Awesome - thanks

Reply:
Have added more tests, think I have covered most cases now.